R/process_source.R

```diff
@@ -1,131 +1,71 @@
-        # starts here.
-        state <- "OBJECT_DEFINITION"
-        current_chunk$add(line)
-        current_chunk_name <- extract_object_name(line)
-
-        # Special case: check if it's a one-liner (one-liners will be parseable;
-        # an opening line for multi-line functions/expressions will not.). If
-        # so, transition to SCANNING state.
-        if (is_parseable(line)) {
-          state <- "SCANNING"
-          chunks[[current_chunk_name]] <- current_chunk$get()
-          current_chunk$reset()
-          current_chunk_name <- NULL
-        }
-
-      } else if (grepl("^#", line)) {
-        # Stay in current state.
-        current_chunk$add(line)
-
-      } else {
-        # We've hit the end of a comment block but no definition here. Just
-        # discard the block and go back to scanning mode.
-        state <- "SCANNING"
-        current_chunk$reset()
-      }
-    }
+# Given a file, process the text and return a named list of objects.
+process_source_file_one <- function(file) {
+  parse_data <- utils::getParseData(parse(file = file, keep.source = TRUE))
+  text <- readLines(file)
+
+  assignment_ops <- parse_data[
+    parse_data$token %in% c("LEFT_ASSIGN", "EQ_ASSIGN", "RIGHT_ASSIGN"),
+  ]
+  # Assignments are the parent expressions of assignment operators
+  assignments <- parse_data[parse_data$id %in% assignment_ops$parent, ]
+  # Top-level expressions have a parent attribute of 0
+  object_definitions <- assignments[assignments$parent == 0, ]
+
+  staticexports <- vector("list", nrow(object_definitions))
+
+  for (i in seq_len(nrow(object_definitions))) {
+    object_definition <- object_definitions[i, ]
+
+    names(staticexports)[i] <- extract_object_name(
+      parse_data, object_definition, assignment_ops
+    )
+
+    # A leading comment's parent attribute is 0 - the next expression's id
+    leading_comments <- parse_data[parse_data$parent == -object_definition$id, ]
+
+    # Exclude leading comments if there are any empty lines between the comment
+    # and the object definition
+    leading_comments <- leading_comments[
+      object_definition$line1 - leading_comments$line1 <=
+        rev(seq_len(nrow(leading_comments))),
+    ]
+
+    staticexport_lines <- seq(
+      min(leading_comments$line1, object_definition$line1),
+      max(leading_comments$line2, object_definition$line2)
+    )
+    staticexport_text <- text[staticexport_lines]
+    staticexports[[i]] <- staticexport_text
   }
 
-  chunks
-}
-
-
-# A wrapper object for vectors, which makes it easy to efficiently append items
-# to it.
-vec_accumulator <- function(init = logical(0)) {
-  s <- init
-  list(
-    add = function(x) {
-      s[length(s) + seq_along(x)] <<- x
-    },
-    get = function() {
-      s
-    },
-    reset = function() {
-      s <<- logical(0)
-    }
-  )
+  staticexports
 }
 
+extract_object_name <- function(parse_data, object_definition, assignment_ops) {
```

Comment:
It would be clearer to pass in just the `object_id` instead of `object_definition`.

Reply:
Good point! Updated in [47370ae](https://github.com/wch/staticimports/pull/21/commits/47370ae48ac2d72d079fbff2c2111a741a4a5c78)